Add Pi Zero 2 W

acostach opened the pull request:
**- What I did**: Added Pi Zero 2 Wireless machines and firmware package

**- How I did it**:  Used the existing Pi3 machines configurations

Connects-to: https://github.com/agherzan/meta-raspberrypi/issues/929

acostach commented:
Hi @agherzan, should we switch to inheriting the pi3 machines in the pi0-2w instead? Also, should there be both 32 and 64 bit versions?

agherzan commented:
@acostach I'm fine with providing both but @kraj has a good point. The new machine only differ in dtb and Bluetooth firmware, right? So I'd say include the respective rpi3 ones and tweak the changes.

agherzan commented on recipes-kernel/linux-firmware-rpidistro/linux-firmware-rpidistro_git.bb:
```diff
@@ -57,7 +57,7 @@ do_install() {
     # raspbian firmware-nonfree. Raspbian adds blobs and nvram
     # definitions that are also necessary so copy those too.
-    for fw in brcmfmac43430-sdio brcmfmac43455-sdio brcmfmac43456-sdio ; do
+    for fw in brcmfmac43430-sdio brcmfmac43436-sdio brcmfmac43455-sdio brcmfmac43456-sdio ; do
```
I'm pretty sure that these new boards were supposed to work with `brcmfmac43436s`. Can you test that and package them as well? You will need a package update for this.

kraj reviewed (dismissed):
lgtm.

acostach commented:
Thanks for reviewing this @agherzan @kraj , I've updated the branch according to your suggestions.
The Pi we have seems to only work with the bcm43436-sdio firmware so I've included both and updated the kernel to 5.10.78 to include the dts wifi changes https://github.com/raspberrypi/linux/commit/f71d77d11ea3a0cc391030b6c4eb5a0b22e9e787

agherzan commented on conf/machine/raspberrypi0-2w-64.conf:
```diff
@@ -0,0 +1,21 @@
+#@NAME: RaspberryPi0 2 Wifi Development Board
+#@DESCRIPTION: Machine configuration for the RaspberryPi0 2 Wifi in 64 bits mode
+
+MACHINEOVERRIDES = "raspberrypi3-64:${MACHINE}"
```
Shall we define this after the include? Just to be safe.

acostach replied:
done

agherzan commented on conf/machine/raspberrypi0-2w.conf:
```diff
@@ -0,0 +1,17 @@
+#@NAME: RaspberryPi0 2 Wifi Development Board (32bit)
+#@DESCRIPTION: Machine configuration for the RaspberryPi0 2 Wifi in 32 bits mode
+
+MACHINEOVERRIDES = "raspberrypi3:${MACHINE}"
```
Shall we define this after the include? Just to be safe.

acostach replied:
done

agherzan commented on conf/machine/raspberrypi0-2w-64.conf:
```diff
@@ -0,0 +1,21 @@
+    broadcom/bcm2710-rpi-zero-2.dtb \
+    "
+
+# This is necessary because the Pi Zero 2 Wifi 64bit overrides the Pi3 64bit,
```
Isn't this because the include overwrites your `MACHINEOVERRIDES`?

acostach replied:
Thanks @agherzan, indeed, the include does replace it and this is how it looks like without the last assignment on it:

MACHINEOVERRIDES="rpi:aarch64:armv8a:raspberrypi3:raspberrypi0-2w-64"

If we use `MACHINEOVERRIDES="raspberrypi3-64:${MACHINE}"` after the include then recipes
will complain about raspberrypi0-2w-64 not present in COMPATIBLE_MACHINE, since rpi is not in MACHINEOVERRIDES.

I modified the PR to do the replace after the include as you suggested, alternatively we could hardcode MACHINEOVERRIDES:
```
MACHINEOVERRIDES = "rpi:armv7ve:raspberrypi3:${MACHINE}"
MACHINEOVERRIDES = "rpi:aarch64:armv8a:raspberrypi3:raspberrypi3-64:${MACHINE}"
```

kraj replied:
hardcoding is not a good idea.

acostach replied:
Sure @kraj will leave the overrides as they are, if it's all good should we merge this?

kraj replied:
yes

agherzan commented on conf/machine/raspberrypi0-2w-64.conf:
```diff
@@ -0,0 +1,17 @@
+
+include conf/machine/raspberrypi3-64.conf
+
+MACHINEOVERRIDES := "${@'${MACHINEOVERRIDES}'.replace(':${MACHINE}',':raspberrypi3-64:${MACHINE}')}"
```
I find this replace just superfluous. Isn't `MACHINEOVERRIDES =. "raspberrypi3-64:"` enough?

agherzan replied:
No. It won't work because the order will be wrong.

agherzan replied:
Let's keep it as such for now.

agherzan commented on conf/machine/raspberrypi0-2w.conf:
```diff
@@ -0,0 +1,17 @@
+
+include conf/machine/raspberrypi3.conf
+
+MACHINEOVERRIDES := "${@'${MACHINEOVERRIDES}'.replace(':${MACHINE}',':raspberrypi3:${MACHINE}')}"
```
I find this replace just superfluous. Isn't `MACHINEOVERRIDES =. "raspberrypi3:"` enough?

agherzan replied:
No. It won't work because the order will be wrong.

agherzan replied:
Let's keep it as such for now.

agherzan commented on conf/machine/raspberrypi0-2w.conf:
```diff
@@ -0,0 +1,17 @@
+"
+
+RPI_KERNEL_DEVICETREE = " \
+    broadcom/bcm2710-rpi-zero-2.dtb \
```
This won't work on 32 bit.

acostach replied:
Ok, I removed it.

agherzan replied:
Have you tried to build the 32 machine?  I'm wondering if the default value of `RPI_KERNEL_DEVICETREE` makes any sense anymore because the arm and arm64 have now different dtbs including different path.

acostach replied:
No, we're only interested in the 64bit image

agherzan replied:
If you only have only tested the 64 variant, I'd say to only add that for now. I'm not sure if the 32 one builds now.

acostach replied:
Sure, I removed it.

agherzan commented:
I'll give it a spin tonight and merge if everything looks good.

agherzan commented:
@agherzan-ci ok to test

acostach commented:
Perfect, thank you Andrei

agherzan reviewed (approved):
Boots fine and wifi looks also fine. Sadly Weston on DRM backend doesn't seem to start.

floion commented:
Thanks @agherzan for merging this. Can it be backported to honister also?

agherzan commented:
Sure @floion . I'll handle this soon.

agherzan commented:
On a second check, it is already in honister. Is there something I'm missing?
